src/net/protocol_agnostic/addr.rs

```diff
@@ -1,6 +1,91 @@
 use priv_prelude::*;
+use url::{self, Url};
+use std::fmt;
+use std::str::FromStr;
+use std::net;
 
 pub enum PaAddr {
     Tcp(SocketAddr),
 }
 
+impl fmt::Display for PaAddr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            PaAddr::Tcp(ref addr) => {
+                write!(f, "tcp://{}", addr)
+            },
+        }
+    }
+}
+
+quick_error! {
+    #[derive(Debug)]
+    pub enum ParseError {
+        MalformedUrl(e: url::ParseError) {
+            description("malformed url")
+            display("malformed url. {}", e)
+            cause(e)
+        }
+        MalformedHost(e: net::AddrParseError) {
+            description("malformed host")
+            display("malformed host. {}", e)
+            cause(e)
+        }
+        UnknownScheme(s: String) {
+            description("unknown scheme")
+            display("unknown scheme: \"{}\"", s)
+        }
+        MissingPort {
+            description("missing port number")
+        }
+    }
+}
+
+impl FromStr for PaAddr {
+    type Err = ParseError;
+
+    fn from_str(s: &str) -> Result<PaAddr, ParseError> {
+        let url = match Url::parse(s) {
+            Ok(url) => url,
+            Err(e) => return Err(ParseError::MalformedUrl(e)),
+        };
+        let ret = match url.scheme() {
+            "tcp" => {
+                let ip = match url.host_str() {
+                    None => return Err(ParseError::MalformedUrl(url::ParseError::EmptyHost)),
+                    Some(host) => match IpAddr::from_str(host) {
+                        Err(e) => return Err(ParseError::MalformedHost(e)),
+                        Ok(addr) => addr,
+                    },
+                };
+                let port = match url.port() {
+                    Some(port) => port,
+                    None => return Err(ParseError::MissingPort),
+                };
+                let addr = SocketAddr::new(ip, port);
+                PaAddr::Tcp(addr)
+            },
+            scheme => return Err(ParseError::UnknownScheme(scheme.to_owned())),
+        };
+        Ok(ret)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use priv_prelude::*;
+    use std::str::FromStr;
+
+    #[test]
+    fn test() {
```

Comment:
Test case description would be nice: in case it fails, it would provide more details what's wrong, .e.g.
```
#[test]
fn from_str_parses_address_from_url()
```